### DISCUSSION TITLE
Requests in iframe count as first-party

### DISCUSSION THREAD

**Discussion by MasterKia** [uBlockOrigin/uAssets#14705] [Q&A] [answered]
First add this to avoid EasyList false positive:
```adb
.com/ads/$subdocument,domain=~developers.google.com|~sabotage-films.com,badfilter
```

Then add this and go to `ejiga.com/w/SQxMqtz`:
```adb
||tamasha.com^$3p,~frame
```

Notice that only a handful of requests except the iframe are blocked:
![image](https://user-images.githubusercontent.com/17685483/188658588-d8f74532-210a-42a3-a327-5f10ec6552f7.png)

Change the last filter to:
```adb
||tamasha.com^$1p,~frame
```

Notice that all requests except the iframe are blocked (as intended):
![Screenshot_20220906_184658](https://user-images.githubusercontent.com/17685483/188658892-86a49b45-c046-4735-92a1-573494242343.png)

But now, `tamasha.com/v/j5B3k` is broken.

The goal is to block `tamasha.com` on **_all_** third-party sites but only allowing its video player assets (using exception filters), but it seems to be impossible without breaking the site itself.

This is impossible to do using uBO because:
> [Any request from an iframe to the iframe URL is 100% 1p.](https://github.com/DandelionSprout/adfilt/issues/63#issuecomment-1238106842)

What can be done about this?

**Comment by Yuki2718** [comment]
> The goal is to block tamasha.com on all third-party sites but only allowing its video player assets 

is already 100% achieved by `||tamasha.com^$3p,~frame`, it's not uBO specifically counts these requests as 1p, but they are in all senses 1p because they come from `tamasha.com` and go to `tamasha.com`.

Simple question: why you wanna block these requests except when user visited the site,  in other words, what specific problem does that solve? I don't understant why
```adb
||tamasha.com/dist/*/contrib/player/js/videojs.ima.min.js
||tamasha.com/dist/*/contrib/player/js/videojs-contrib.min.js
||tamasha.com/dist/*/contrib/player/js/ima
```

is not desirable.

**Reply by MasterKia** [reply]
> is already 100% achieved by `||tamasha.com^$3p,~frame`

A lot of requests are **not** blocked as you can see in the first screenshot.

> Simple question

The site offers embedded video player.
Instead of blocking the ads/tracker scripts (which can change from time to time, e.g. the site has 5 variations of ads/tracker scripts as of now), my goal is to completely block the third-party requests, then add exceptions for the video player assets (which don't change from time to time, i.e. a video player is still a video player after 2 months).

**Reply by Yuki2718** [reply]
This is becoming a loop, you somehow think 1p requests should be treated as 3p.

**Reply by MasterKia** [reply]
> it's not that uBO specifically counts these requests as 1p, but they are in all senses 1p.

In this context I don't think they qualify as 1p requests.

ejiga.com => tamasha.com [iframe] => tamasha.com [scripts]

Why count the `[scripts]` request as 1p when it's initiated by ejiga.com loading the tamasha.com iframe in the first place?

**Reply by Yuki2718** [reply]
I give up and let others to explain.

**Reply by MasterKia** [reply]
If ejiga.com is removed from the equation, **_then_** it should be count as 1p request.

**Reply by MasterKia** [reply]
> I don't understant why these filters are not desirable.

Because Ads/tracking scripts change faster than video player scripts.

**Comment by gorhill** [comment, marked as answer]
> What can be done about this?

Nothing, it's the actual semantic of 1p/3p in easylist-compatible filter syntax -- changing the semantic would lead to widespread breakage. That's just the way things work.

**Reply by MasterKia** [reply]
What about changing the semantics to count requests as 3p **_only_** if they're initiated by an iframe?

How would that lead to widespread breakage?
(Sorry I can't come up with any real world example).

**Reply by Yuki2718** [reply]
ANY 1p exception won't work for frame in that case. E.g. `@@||youtube.com/api/analytics/$~third-party` won't work for embedded player and if it's still needed all embedded YT will be broken. There are no difference btwn iframe (subdocument) and main document, virtually any websites can be embedded as iframe. IMHO you have to overcome your desire for what you think things should be, and accept and get yourself to be used to what things are and logic behind them (https://github.com/uBlockOrigin/uBlock-issues/issues/2244 and some past comments are in my mind).

**Reply by MasterKia** [reply]
> you have to overcome your desire for what you think things should be
> accept and get yourself to be used to what things are

That's not how new features are born, e.g. this would've never been possible if the user forced himself to use cosmetic filters:
https://github.com/gorhill/uBlock/commit/152120bd9ec7a2ccea907e015fb195484ef7bc8e

**Reply by Yuki2718** [reply]
New feature born by understanding and following current logic of how web and uBO work and still finding something not filled by current implementation.

**Comment by Yuki2718** [comment]
@gorhill Is it technically possible to take root context into account? A problem is `@@||tpc.googlesyndication.com/pagead/js/$domain=googleads.g.doubleclick.net` in EL is needed for `https://game.goo.ne.jp/goo/game/469` but not for `https://anydebrid.com/` or `https://skmedix.pl/downloads`.

**Reply by uBlock-user** [reply]
Should be. It's already done for dynamic filtering.

**Reply by gorhill** [reply]
> technically possible to take root context into account?

We discussed a `rootDomain` option somewhere else a long time ago. Here is not the place for such discussion, as it's unrelated and there is no way I will ever change the semantic of `1p`/`3p` in uBO's static network filtering engine, which is what the discussion was about here.

**Reply by uBlock-user** [reply]
@gorhill I suggested rootdomain in https://github.com/uBlockOrigin/uBlock-discussions/discussions/354#discussioncomment-5307255. Are you still open to implementing it ?

**Reply by gorhill** [reply]
That's the one I was thinking about. As usual, it's about the balance of work required/future maintenance of that code vs. number of cases solved.

**Reply by uBlock-user** [reply]
It will solve mine and this case alongwith all similar cases where iframes is involved.

**Reply by stephenhawk8054** [reply]
This issue might also need the `root` solution: https://github.com/uBlockOrigin/uAssets/issues/16014#issuecomment-1732295421
